File: src/test/java/io/pivotal/project/ProjectServiceTest.java
package io.pivotal.project;

import io.pivotal.project.burndown.BurndownCsvParser;
import io.pivotal.project.burndown.BurndownPeriod;
import org.junit.Before;
import org.junit.Test;
import org.mockito.ArgumentCaptor;

import java.time.LocalDate;
import java.util.*;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Matchers.any;
import static org.mockito.Matchers.anyInt;
import static org.mockito.Mockito.*;

public class ProjectServiceTest {

    private ProjectRepository mockProjectRepository;
    private BurndownCsvParser mockBurndownCsvParser;
    private ProjectService projectService;

    @Before
    public void setup() {
        mockProjectRepository = mock(ProjectRepository.class);
        mockBurndownCsvParser = mock(BurndownCsvParser.class);

        projectService = new ProjectService(mockProjectRepository, mockBurndownCsvParser);
    }

    @Test
    public void getAllProjects_setsProjectFieldsFromRepositoryEntities() {
        ProjectEntity firstProjectEntity = new ProjectEntity(123, "project-1", LocalDate.MAX, 500, 10000);
        ProjectEntity secondProjectEntity = new ProjectEntity(456, "project-2", LocalDate.MIN, 250, 75000);
        List<ProjectEntity> projectEntities = Arrays.asList(firstProjectEntity, secondProjectEntity);
        when(mockProjectRepository.getAllProjectEntities()).thenReturn(projectEntities);

        List<Project> allProjects = projectService.getAllProjects();

        assertThat(allProjects).hasSize(2);
        Project firstProject = allProjects.get(0);
        assertThat(firstProject.getId()).isEqualTo(123);
        assertThat(firstProject.getName()).isEqualTo("project-1");
        assertThat(firstProject.getHourlyRate()).isEqualTo(500);
        assertThat(firstProject.getStartDate()).isEqualTo(LocalDate.MAX);
        assertThat(firstProject.getBudget()).isEqualTo(10000);

        Project secondProject = allProjects.get(1);
        assertThat(secondProject.getId()).isEqualTo(456);
        assertThat(secondProject.getName()).isEqualTo("project-2");
        assertThat(secondProject.getHourlyRate()).isEqualTo(250);
        assertThat(secondProject.getStartDate()).isEqualTo(LocalDate.MIN);
        assertThat(secondProject.getBudget()).isEqualTo(75000);
    }

    @Test
    public void getProjectById_setsProjectFieldsFromRepositoryEntity() {
        LocalDate stubbedLocalDate = LocalDate.of(2017, 2, 13);
        int stubbedHourlyRate = 500;
        ProjectEntity stubbedProjectEntity = new ProjectEntity(123, "foo-bar", stubbedLocalDate, stubbedHourlyRate, 50000);
        when(mockProjectRepository.getProjectEntityById(56)).thenReturn(stubbedProjectEntity);

        Optional<Project> project = projectService.getProjectById(56);

        Project actualProject = project.get();
        assertThat(actualProject.getId()).isEqualTo(123);
        assertThat(actualProject.getName()).isEqualTo("foo-bar");
        assertThat(actualProject.getStartDate()).isEqualTo(stubbedLocalDate);
        assertThat(actualProject.getHourlyRate()).isEqualTo(stubbedHourlyRate);
        assertThat(actualProject.getBudget()).isEqualTo(50000);
    }

    @Test
    public void getProjectById_setsBurndownListFromCsvParser() {
        ProjectEntity stubbedProjectEntity = new ProjectEntity(2, "some-name-that-matches-openair", null, 0, 0);
        when(mockProjectRepository.getProjectEntityById(100)).thenReturn(stubbedProjectEntity);

        ArrayList<BurndownPeriod> stubbedBurndown = new ArrayList<>();
        when(mockBurndownCsvParser.getBurndownForProjectEntity(stubbedProjectEntity)).thenReturn(stubbedBurndown);

        Optional<Project> project = projectService.getProjectById(100);

        Project actualProject = project.get();
        assertThat(actualProject.getBurndown()).isSameAs(stubbedBurndown);
    }

    @Test
    public void getProjectById_defaultsBurndownToAnEmptyList() {
        when(mockProjectRepository.getProjectEntityById(anyInt())).thenReturn(new ProjectEntity());
        when(mockBurndownCsvParser.getBurndownForProjectEntity(any(ProjectEntity.class))).thenReturn(null);

        Project project = projectService.getProjectById(new Random().nextInt()).get();

        assertThat(project.getBurndown()).isEmpty();
    }

    @Test
    public void getProjectById_returnsEmptyOptionalWhenRepositoryReturnsNull() {
        when(mockProjectRepository.getProjectEntityById(anyInt())).thenReturn(null);

        Optional<Project> projectById = projectService.getProjectById(new Random().nextInt());

        assertThat(projectById.isPresent()).isFalse();
    }

    @Test
    public void saveProject_delegatesToProjectRepository() {
        when(mockProjectRepository.save(any())).thenReturn(new ProjectEntity());

        Project project = new Project();
        project.setName("project-name");
        project.setHourlyRate(150);
        project.setStartDate(LocalDate.MAX);
        project.setBudget(75000);
        projectService.saveProject(project);

        ArgumentCaptor<ProjectEntity> captor = ArgumentCaptor.forClass(ProjectEntity.class);
        verify(mockProjectRepository).save(captor.capture());
        ProjectEntity value = captor.getValue();
        assertThat(value.getName()).isEqualTo("project-name");
        assertThat(value.getHourlyRate()).isEqualTo(150);
        assertThat(value.getStartDate()).isEqualTo(LocalDate.MAX);
        assertThat(value.getBudget()).isEqualTo(75000);
    }

    @Test
    public void saveProject_returnsNewlyCreatedProject() {
        ProjectEntity stubbedProject = new ProjectEntity(10, "baz", LocalDate.MIN, 75, 10000);
        when(mockProjectRepository.save(any())).thenReturn(stubbedProject);

        Project createdProject = projectService.saveProject(new Project());

        assertThat(createdProject.getId()).isEqualTo(10);
        assertThat(createdProject.getName()).isEqualTo("baz");
        assertThat(createdProject.getHourlyRate()).isEqualTo(75);
        assertThat(createdProject.getStartDate()).isEqualTo(LocalDate.MIN);
        assertThat(createdProject.getBudget()).isEqualTo(10000);
        assertThat(createdProject.getBurndown()).isEmpty();
    }

}

File: src/test/java/io/pivotal/project/ProjectServiceTest.kt
package io.pivotal.project

import io.pivotal.project.burndown.BurndownParser
import io.pivotal.project.testhelpers.aProject
import io.pivotal.project.testhelpers.aProjectEntity
import org.junit.Test
import org.mockito.ArgumentCaptor

import java.time.LocalDate
import java.util.*

import org.assertj.core.api.Assertions.assertThat
import org.mockito.Matchers.any
import org.mockito.Matchers.anyInt
import org.mockito.Mockito
import org.mockito.Mockito.*

class ProjectServiceTest {

    private val mockProjectRepository = mock<ProjectRepository>(ProjectRepository::class.java)
    private val mockBurndownParser = Mockito.mock<BurndownParser>(BurndownParser::class.java)
    private val projectService: ProjectService = ProjectService(mockProjectRepository, mockBurndownParser)

    @Test
    fun getAllProjects_setsProjectFieldsFromRepositoryEntities() {
        val firstProjectEntity = ProjectEntity(123, "project-1", LocalDate.MAX, 500, 10000)
        val secondProjectEntity = ProjectEntity(456, "project-2", LocalDate.MIN, 250, 75000)
        val projectEntities = Arrays.asList(firstProjectEntity, secondProjectEntity)
        `when`(mockProjectRepository.allProjectEntities()).thenReturn(projectEntities)

        val allProjects = projectService.allProjects()

        assertThat(allProjects).hasSize(2)
        val firstProject = allProjects[0]
        assertThat(firstProject.id).isEqualTo(123)
        assertThat(firstProject.name).isEqualTo("project-1")
        assertThat(firstProject.hourlyRate).isEqualTo(500)
        assertThat(firstProject.startDate).isEqualTo(LocalDate.MAX)
        assertThat(firstProject.budget).isEqualTo(10000)

        val secondProject = allProjects[1]
        assertThat(secondProject.id).isEqualTo(456)
        assertThat(secondProject.name).isEqualTo("project-2")
        assertThat(secondProject.hourlyRate).isEqualTo(250)
        assertThat(secondProject.startDate).isEqualTo(LocalDate.MIN)
        assertThat(secondProject.budget).isEqualTo(75000)
    }

    @Test
    fun getProjectById_setsProjectFieldsFromRepositoryEntity() {
        val stubbedLocalDate = LocalDate.of(2017, 2, 13)
        val stubbedHourlyRate = 500
        val stubbedProjectEntity = ProjectEntity(123, "foo-bar", stubbedLocalDate, stubbedHourlyRate, 50000)
        `when`(mockProjectRepository.getProjectEntityById(56)).thenReturn(stubbedProjectEntity)

        val project = projectService.getProjectById(56)

        assertThat(project!!.id).isEqualTo(123)
        assertThat(project.name).isEqualTo("foo-bar")
        assertThat(project.startDate).isEqualTo(stubbedLocalDate)
        assertThat(project.hourlyRate).isEqualTo(stubbedHourlyRate)
        assertThat(project.budget).isEqualTo(50000)
    }

    @Test
    fun getProjectById_setsBurndownListFromCsvParser() {
        val stubbedProjectEntity = aProjectEntity()
        `when`(mockProjectRepository.getProjectEntityById(100)).thenReturn(stubbedProjectEntity)

        val stubbedBurndown = ArrayList<Float>()
        `when`(mockBurndownParser.getBurndownForProjectEntity(stubbedProjectEntity)).thenReturn(stubbedBurndown)

        val project = projectService.getProjectById(100)

        assertThat(project!!.burndown).isSameAs(stubbedBurndown)
    }

    @Test
    fun getProjectById_defaultsBurndownToAnEmptyList() {
        `when`(mockProjectRepository.getProjectEntityById(anyInt())).thenReturn(aProjectEntity())
        `when`(mockBurndownParser.getBurndownForProjectEntity(any<ProjectEntity>(ProjectEntity::class.java))).thenReturn(null)

        val project = projectService.getProjectById(Random().nextInt())

        assertThat(project!!.burndown).isEmpty()
    }

    @Test
    fun getProjectById_returnsEmptyOptionalWhenRepositoryReturnsNull() {
        `when`(mockProjectRepository.getProjectEntityById(anyInt())).thenReturn(null)

        val projectById = projectService.getProjectById(Random().nextInt())

        assertThat(projectById).isNull()
    }

    @Test
    fun saveProject_delegatesToProjectRepository() {
        `when`(mockProjectRepository.save(any<ProjectEntity>())).thenReturn(aProjectEntity())

        val project = Project(
                id = null,
                name = "project-name",
                hourlyRate = 150,
                startDate = LocalDate.MAX,
                budget = 75000
        )
        projectService.saveProject(project)

        val captor = ArgumentCaptor.forClass<ProjectEntity>(ProjectEntity::class.java)
        verify<ProjectRepository>(mockProjectRepository).save(captor.capture())
        val value = captor.value
        assertThat(value.name).isEqualTo("project-name")
        assertThat(value.hourlyRate).isEqualTo(150)
        assertThat(value.startDate).isEqualTo(LocalDate.MAX)
        assertThat(value.budget).isEqualTo(75000)
    }

    @Test
    fun saveProject_returnsNewlyCreatedProject() {
        val stubbedProject = ProjectEntity(10, "baz", LocalDate.MIN, 75, 10000)
        `when`(mockProjectRepository.save(any<ProjectEntity>())).thenReturn(stubbedProject)

        val createdProject = projectService.saveProject(aProject())

        assertThat(createdProject.id).isEqualTo(10)
        assertThat(createdProject.name).isEqualTo("baz")
        assertThat(createdProject.hourlyRate).isEqualTo(75)
        assertThat(createdProject.startDate).isEqualTo(LocalDate.MIN)
        assertThat(createdProject.budget).isEqualTo(10000)
        assertThat(createdProject.burndown).isEmpty()
    }


}